### PR TITLE
[devscripts] add cert check and graceful shutdown

### PR DESCRIPTION
**Pull request by psathyan**
Currently, the OpenShift cluster is abruptly shutdown before preparing for the golden image. This could lead to the keepalived pod not starting and/or having expired API certificates.

As part of this change set, the following tasks are added
- Splitting `110_check_ocp.yml` into two
- - `111_verify_cluster.yml` contains the login check
- - `112_verify_golden_image.yml` contains the API certificate check
- Call to graceful shutdown as provided by `openshift_adm` role before initiating preparation of the overlay.

_Additional changes_
- `post stage` now should wait for a stable cluster
- Certificate expiration date is compared with today's date and if we have two hours gap... the cluster is brought up.

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running

Depends-On: https://github.com/openstack-k8s-operators/ci-framework/pull/1213
Closes: [OSPRH-5115](https://issues.redhat.com//browse/OSPRH-5115)

*Testing Results*

_Rerun scenario_
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=363  changed=126  unreachable=0    failed=0    skipped=160  rescued=0    ignored=0   

Tuesday 05 March 2024  06:36:10 -0500 (0:00:00.172)       0:15:19.135 ********* 
=============================================================================== 
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.14s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 50.71s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 50.58s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 50.37s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 49.07s
reproducer : Install some tools ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 37.73s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 26.80s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 26.55s
libvirt_manager : Grab IPs for nodes type compute ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 20.19s
networking_mapper : Ensure that networking and hostname facts are in place ----------------------------------------------------------------------------------------------------------------------------------------------------------- 19.84s
reproducer : Inject ProxyJump configuration for remote hypervisor VMs ---------------------------------------------------------------------------------------------------------------------------------------------------------------- 19.70s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.94s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.84s
libvirt_manager : Wait for SSH on VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.89s
libvirt_manager : Prepare disk image with SSH and growing volume --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 16.75s
libvirt_manager : Configure VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 13.69s
reproducer : Wait for OCP nodes to be ready ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 12.47s
reproducer : Install ansible dependencies -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 12.46s
libvirt_manager : Download base image ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 11.21s
reproducer : Inject kubeconfig content ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 6.56s
```

_Greenfield install_
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=405  changed=168  unreachable=0    failed=0    skipped=172  rescued=2    ignored=0   

Tuesday 05 March 2024  20:25:38 -0500 (0:00:00.223)       1:12:18.456 ********* 
=============================================================================== 
devscripts : Deploy OpenShift cluster ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 3037.39s
reproducer : Install internal CA ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 152.38s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 79.88s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 78.90s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 78.24s
reproducer : Install some tools ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 46.93s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 32.28s
reproducer : Inject kubeadmin-password if exists ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 31.95s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 28.37s
reproducer : Inject devscripts private key if set ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 26.90s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 25.16s
networking_mapper : Ensure that networking and hostname facts are in place ----------------------------------------------------------------------------------------------------------------------------------------------------------- 23.57s
libvirt_manager : Prepare disk image with SSH and growing volume --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 20.47s
libvirt_manager : Grab IPs for nodes type compute ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 20.44s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 19.04s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.71s
libvirt_manager : Wait for SSH on VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.93s
reproducer : Wait for OCP nodes to be ready ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 13.62s
libvirt_manager : Configure VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 12.94s
reproducer : Install ansible dependencies -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 12.80s
[ciuser@devci-01 ci-framework]$ 

[ciuser@titanamd126 ~]$ oc get nodes
NAME       STATUS                     ROLES                         AGE   VERSION
master-0   Ready,SchedulingDisabled   control-plane,master,worker   84m   v1.28.6+6216ea1
master-1   Ready,SchedulingDisabled   control-plane,master,worker   84m   v1.28.6+6216ea1
master-2   Ready,SchedulingDisabled   control-plane,master,worker   84m   v1.28.6+6216ea1
```

_Rerun 2_
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=370  changed=128  unreachable=0    failed=0    skipped=160  rescued=0    ignored=0   

Wednesday 06 March 2024  02:42:43 -0500 (0:00:00.138)       0:29:30.895 ******* 
=============================================================================== 
openshift_adm : Wait unit the OpenShift cluster is stable. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 595.33s
openshift_adm : Wait until OCP login succeeds. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 233.80s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 52.22s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 52.19s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.59s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.12s
reproducer : Install some tools ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 38.99s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 28.44s
openshift_adm : Wait until the OCP API endpoint is reachable. ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 28.31s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 28.27s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 25.74s
networking_mapper : Ensure that networking and hostname facts are in place ----------------------------------------------------------------------------------------------------------------------------------------------------------- 24.28s
libvirt_manager : Grab IPs for nodes type compute ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 20.29s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 19.04s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.70s
libvirt_manager : Wait for SSH on VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.91s
libvirt_manager : Prepare disk image with SSH and growing volume --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 16.68s
libvirt_manager : Configure VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 15.08s
reproducer : Wait for OCP nodes to be ready ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 13.74s
reproducer : Install ansible dependencies -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 13.28s
[ciuser@devci-01 ci-framework]$ 

ciuser@titanamd126 ~]$ oc get nodes
NAME       STATUS   ROLES                         AGE    VERSION
master-0   Ready    control-plane,master,worker   7h9m   v1.28.6+6216ea1
master-1   Ready    control-plane,master,worker   7h9m   v1.28.6+6216ea1
master-2   Ready    control-plane,master,worker   7h9m   v1.28.6+6216ea1
```

### CODE REVIEW

**Comment by psathyan** [comment]
Latest code testing results
```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
titanamd126                : ok=421  changed=152  unreachable=0    failed=0    skipped=168  rescued=3    ignored=0   

Thursday 07 March 2024  08:51:40 -0500 (0:00:00.131)       1:23:24.670 ******** 
=============================================================================== 
devscripts : Deploy OpenShift cluster ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2656.64s
openshift_adm : Wait unit the OpenShift cluster is stable. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 760.81s
openshift_adm : Wait till shutdown is completed on all the nodes -------------------------------------------------------------------------------------------------------------------------------------------------------------------- 300.10s
openshift_adm : Wait unit the OpenShift cluster is stable. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 120.80s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 54.08s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.52s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 51.48s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 50.86s
reproducer : Install some tools ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 40.08s
openshift_adm : Wait until OCP login succeeds. --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 38.17s
networking_mapper : Ensure that networking and hostname facts are in place ----------------------------------------------------------------------------------------------------------------------------------------------------------- 32.02s
openshift_adm : Wait until the OCP API endpoint is reachable. ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 30.89s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 29.34s
openshift_adm : Initiate shutdown ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 27.31s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 26.84s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 26.53s
libvirt_manager : Grab IPs for nodes type compute ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 20.33s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 18.95s
libvirt_manager : Wait for SSH on VMs type compute ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.85s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 17.72s
```

**Comment by psathyan** [comment]
*Latest rerun testing results*
```
TASK [Ensure the OpenShift cluster is stable. name=openshift_adm] ****************************************************************************************************************************************************************************
Thursday 07 March 2024  23:14:56 -0500 (0:00:00.708)       0:06:08.232 ******** 

TASK [openshift_adm : Prepare and shutdown the OpenShift cluster. _raw_params=shutdown.yml] **************************************************************************************************************************************************
Thursday 07 March 2024  23:14:56 -0500 (0:00:00.127)       0:06:08.360 ******** 
skipping: [titanamd126]

TASK [openshift_adm : Wait for the OpenShift cluster to become stable. _raw_params=wait_for_cluster.yml] *************************************************************************************************************************************
Thursday 07 March 2024  23:14:57 -0500 (0:00:00.107)       0:06:08.468 ******** 
included: /home/ciuser/src/ci-framework/roles/openshift_adm/tasks/wait_for_cluster.yml for titanamd126

TASK [openshift_adm : Wait until the OCP API endpoint is reachable. url={{ cifmw_openshift_api }}, return_content=True, validate_certs=False, status_code=403] *******************************************************************************
Thursday 07 March 2024  23:14:57 -0500 (0:00:00.122)       0:06:08.590 ******** 
FAILED - RETRYING: [titanamd126]: Wait until the OCP API endpoint is reachable. (30 retries left).
FAILED - RETRYING: [titanamd126]: Wait until the OCP API endpoint is reachable. (29 retries left).
ok: [titanamd126]

TASK [openshift_adm : Wait until OCP login succeeds. host={{ cifmw_openshift_api }}, password={{ cifmw_openshift_password }}, state=present, username={{ cifmw_openshift_user }}, validate_certs=False] **************************************
Thursday 07 March 2024  23:15:31 -0500 (0:00:34.045)       0:06:42.636 ******** 
FAILED - RETRYING: [titanamd126]: Wait until OCP login succeeds. (30 retries left).
FAILED - RETRYING: [titanamd126]: Wait until OCP login succeeds. (29 retries left).
FAILED - RETRYING: [titanamd126]: Wait until OCP login succeeds. (28 retries left).
FAILED - RETRYING: [titanamd126]: Wait until OCP login succeeds. (27 retries left).
FAILED - RETRYING: [titanamd126]: Wait until OCP login succeeds. (26 retries left).
ok: [titanamd126]

TASK [openshift_adm : Gather the list of nodes kind=Node, kubeconfig={{ cifmw_openshift_adm_kubeconfig }}, validate_certs=False] *************************************************************************************************************
Thursday 07 March 2024  23:16:23 -0500 (0:00:52.102)       0:07:34.738 ******** 
[WARNING]: kubernetes<24.2.0 is not supported or tested. Some features may not work.
ok: [titanamd126]

TASK [openshift_adm : Ensure the nodes are in ready state. name={{ item }}, state=uncordon, kubeconfig={{ cifmw_openshift_adm_kubeconfig }}, validate_certs=False] ***************************************************************************
Thursday 07 March 2024  23:16:26 -0500 (0:00:02.885)       0:07:37.624 ******** 
changed: [titanamd126] => (item=master-0)
changed: [titanamd126] => (item=master-1)
changed: [titanamd126] => (item=master-2)

TASK [openshift_adm : Wait unit the OpenShift cluster is stable. _raw_params=oc adm wait-for-stable-cluster --minimum-stable-period=5s --timeout=15m] ****************************************************************************************
Thursday 07 March 2024  23:16:32 -0500 (0:00:06.580)       0:07:44.205 ******** 
```

**Comment by psathyan** [comment]
*Cleanup on expired certificate*
```
TASK [devscripts : Verify the golden image. _raw_params=112_verify_golden_image.yml] *********************************************************************************************************************************************************
Friday 08 March 2024  05:21:33 -0500 (0:00:00.133)       0:03:58.231 ********** 
included: /home/ciuser/src/ci-framework/roles/devscripts/tasks/112_verify_golden_image.yml for titanamd126

TASK [devscripts : Check for certificate regeneration file. path={{
  (cifmw_devscripts_config.working_dir, '.cert.regenerated') |
  ansible.builtin.path_join
}}] ******************************************************************************
Friday 08 March 2024  05:21:33 -0500 (0:00:00.245)       0:03:58.476 ********** 
ok: [titanamd126]

TASK [devscripts : Read the certificate expiration date recorded. src={{
  (cifmw_devscripts_config.working_dir, '.cert.regenerated') |
  ansible.builtin.path_join
}}] *************************************************************************
Friday 08 March 2024  05:21:34 -0500 (0:00:01.120)       0:03:59.597 ********** 
ok: [titanamd126]

TASK [devscripts : Reset golden image exists flag cifmw_devscripts_ocp_exists=False, cifmw_devscripts_force_cleanup=True] ********************************************************************************************************************
Friday 08 March 2024  05:21:35 -0500 (0:00:00.718)       0:04:00.315 ********** 
ok: [titanamd126]

TASK [devscripts : Cleanup the golden image if the certificate has expired. _raw_params=cleanup.yml] *****************************************************************************************************************************************
Friday 08 March 2024  05:21:35 -0500 (0:00:00.145)       0:04:00.460 ********** 
included: /home/ciuser/src/ci-framework/roles/devscripts/tasks/cleanup.yml for titanamd126

TASK [libvirt_manager : Get installed packages list ] ****************************************************************************************************************************************************************************************
Friday 08 March 2024  05:21:36 -0500 (0:00:00.225)       0:04:00.685 ********** 
ok: [titanamd126]

TASK [libvirt_manager : List all of the existing virtual machines command=list_vms, uri=qemu:///system] **************************************************************************************************************************************
Friday 08 March 2024  05:21:38 -0500 (0:00:02.185)       0:04:02.871 ********** 
ok: [titanamd126]

TASK [libvirt_manager : Get pool configuration command=facts, uri=qemu:///system] ************************************************************************************************************************************************************
Friday 08 March 2024  05:21:39 -0500 (0:00:01.190)       0:04:04.061 ********** 
ok: [titanamd126]

TASK [libvirt_manager : Filter out target environment cleanup_vms={{ vms_list.list_vms | select('match', '^cifmw-.*$') }}] *******************************************************************************************************************
Friday 08 March 2024  05:21:40 -0500 (0:00:00.996)       0:04:05.058 ********** 
ok: [titanamd126]
```

**Comment by psathyan** [comment]
recheck

**Comment by psathyan** [comment]
*Latest run testing results*

```
PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
hypervisor                 : ok=301  changed=92   unreachable=0    failed=0    skipped=138  rescued=1    ignored=0   

Monday 11 March 2024  19:43:17 -0400 (0:00:00.073)       0:11:29.457 ********** 
=============================================================================== 
openshift_adm : Wait until OCP login succeeds. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 176.76s
reproducer : Bootstrap environment on controller-0 ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 64.17s
reproducer : Install some tools ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 53.64s
reproducer : Install internal CA ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 47.06s
libvirt_manager : Grab IPs for nodes type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 43.06s
openshift_adm : Wait until the OCP API endpoint is reachable. ------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 29.65s
libvirt_manager : Prepare disk image with SSH and growing volume --------------------------------------------------------------------------------------------------------------------------------------------------------------------- 24.04s
devscripts : Login to the deployed OpenShift cluster. -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 19.21s
libvirt_manager : Grab IPs for nodes type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 16.10s
libvirt_manager : Wait for SSH on VMs type ocp --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 15.97s
reproducer : Install collections on controller-0 ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 15.89s
reproducer : Install ansible dependencies -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 15.52s
libvirt_manager : Download base image ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 13.64s
openshift_adm : Wait unit the OpenShift cluster is stable. ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 6.60s
networking_mapper : Ensure that networking and hostname facts are in place ------------------------------------------------------------------------------------------------------------------------------------------------------------ 5.65s
libvirt_manager : Wait for SSH on VMs type controller --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 5.33s
reproducer : Wait for OCP nodes to be ready ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 4.92s
reproducer : Get rhos-release --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 4.70s
openshift_adm : Ensure the nodes are in ready state. ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 3.39s
libvirt_manager : Start VMs for type ocp ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 2.94s
[ciuser@devci-01 ci-framework]$ 
```

File: roles/devscripts/tasks/310_prepare_overlay.yml [hunk 1]
```diff
@@ -21,6 +21,18 @@
...
+  vars:
+    cifmw_openshift_adm_op: "shutdown"
+    cifmw_openshift_adm_cert_expire_date_file: >-
+      {{
+        (
+          cifmw_devscripts_config.working_dir, '.ocp_cert_not_after'
+        ) | ansible.builtin.path_join
+      }}
+  ansible.builtin.import_role:
+    name: openshift_adm
```
Comment (cjeanner) [inline roles/devscripts/tasks/310_prepare_overlay.yml:34]: I think this one actually shuts the VM down on its own? Meaning the next task may be useless (`virsh destroy` kind of pull the power cord out)?

Comment (cjeanner) [reply roles/devscripts/tasks/310_prepare_overlay.yml:34]: sounds good! Drained SSH doesn't mean "host down" indeed.

**Review by raukadah — APPROVED** [review]

**Pull request merged**